### PR TITLE
Skipping blackwell tests for sm_110+ architectures.

### PR DESCRIPTION
**Pull request by mdavis36**
Blackwell specific tests use instruction sets for sm_100/104, these tests fail for sm_110+ (like 5090 GPUs).

### CODE REVIEW

**Comment by mdavis36** [comment]
!test

**Comment by mdavis36** [comment]
!test

**Comment by mdavis36** [comment]
!test

**Comment by mdavis36** [comment]
!test

**Comment by mdavis36** [comment]
!test

**Comment by mdavis36** [comment]
!test

File: python/python_frontend/fusion_cache.cpp [hunk 1]
```diff
@@ -347,6 +347,27 @@ FusionCache* FusionCache::get(
     if (load_from_default_workspace && fs::exists(file_path)) {
       try {
         singleton_->deserialize(file_path);
+
+        // Check if deserialized cache exceeds max_fusions limit
+        if (singleton_->fusions_.size() > max_fusions) {
```
Comment (wujingyue) [inline python/python_frontend/fusion_cache.cpp:352]: @rdspring1 I thought FusionCache is no longer used in unit tests. Maybe manual_ci.sh needs an update, e.g., https://github.com/NVIDIA/Fuser/pull/5284/files#diff-40d1427cbdb57c064bd6306c2723a30031e86e5551e1031b8e4f3bb24031364cR34-R35?

File: tests/cpp/test_memory.cpp [hunk 2]
```diff
@@ -558,6 +558,12 @@ TEST_P(TMALoadTestWithABroadcastDim, LoadWithBroadcast) {
   FusionGuard fg(&fusion);
   auto shape = std::get<0>(GetParam());
 
+  size_t required_smem = dataTypeSizeByte(dtype);
```
Comment (wujingyue) [inline tests/cpp/test_memory.cpp]:
```suggestion
  int64_t required_smem = dataTypeSizeByte(dtype);
```

https://google.github.io/styleguide/cppguide.html#Integer_Types

File: tests/cpp/test_combined_inner_outer_reduction.cpp [hunk 3]
```diff
@@ -1076,7 +1076,7 @@ class TmaWarpSpecializedTest
 };
 
 TEST_P(TmaWarpSpecializedTest, SimpleFusion) {
-  NVFUSER_TEST_CUDA_ARCH_GUARD(9, 0);
+  NVFUSER_TEST_CUDA_ARCH_RANGE_GUARD(9, 0, 11, 0);
```
Comment (naoyam) [inline tests/cpp/test_combined_inner_outer_reduction.cpp]: What features are missing for this test?

File: tests/python/direct/test_repro.py [hunk 4]
```diff
@@ -6,6 +6,7 @@
 import torch
 import pytest
 from nvfuser_direct import FusionDefinition, DataType
+from conftest import require_device_mem_size_gb
```
Comment (wujingyue) [inline tests/python/direct/test_repro.py]: conftest is for pytest fixtures, not regular testing utilities. 

Consider `from python.direct_utils`

File: tests/python/direct/conftest.py [hunk 5]
```diff
@@ -12,6 +12,19 @@
 from python.direct_utils import is_pre_volta, check_captured_python_definition
 
 
+def require_device_mem_size_gb(required_gb=32.0, didx=0):
```
Comment (wujingyue) [inline tests/python/direct/conftest.py]:
```suggestion
def skip_if_global_memory_below_gb(min_gb: int, gpu_id: int = 0):
```
`skip_if` instead of `require` because that more accurately describes the function. `require` could mean an assertion failure. 

`global` instead of `device` for accuracy. There's also shared memory size. 

`memory` instead of `mem` for https://google.github.io/styleguide/cppguide.html#General_Naming_Rules and consistency since your implementation uses `memory` already.

File: tests/python/direct/conftest.py [hunk 6]
```diff
@@ -12,6 +12,19 @@
...
+    # Get properties for the default CUDA device (GPU 0)
+    device_properties = torch.cuda.get_device_properties(didx)
+    total_memory_bytes = device_properties.total_memory
+    total_memory_gb = total_memory_bytes / (1024**3)
```
Comment (wujingyue) [inline tests/python/direct/conftest.py]: I'd convert gbs to bytes for precision. Floating point operations are lossy sometimes.

**Review by wujingyue — APPROVED** [review]

File: tests/python/direct/test_cutlass_nvfp4_gemm.py [hunk 7]
```diff
@@ -7,9 +7,10 @@
 import torch
 from nvfuser_direct import nvf_cutlass
 
-if torch.cuda.get_device_capability() < (10, 0):
+compute_cap = torch.cuda.get_device_capability()
+if compute_cap < (10, 0) or compute_cap >= (12, 0):
```
Comment (rdspring1) [inline tests/python/direct/test_cutlass_nvfp4_gemm.py:11]: Will this work for GB300? I think its compute capability is `10.3`.

File: tests/python/direct/test_cutlass_gemm.py [hunk 8]
```diff
@@ -6,12 +6,16 @@
...
 from python.direct_utils import is_pre_blackwell
+from python.direct_utils import is_pre_blackwell_12
 from nvfuser_direct import nvf_cutlass
 
 
 @pytest.mark.skipif(
     is_pre_blackwell(), reason="Only supported on blackwell and newer devices."
 )
+@pytest.mark.skipif(
+    not is_pre_blackwell_12(), reason="Only supported on blackwell and newer devices."
```
Comment (rdspring1) [inline tests/python/direct/test_cutlass_gemm.py]:
```suggestion
    not is_pre_blackwell_12(), reason="Does not support blackwell compute 12.0."
```

File: python/python_frontend/fusion_cache.cpp (hunk 1 above)
Comment (rdspring1) [reply python/python_frontend/fusion_cache.cpp:352]: The pytest `tests/python/test_python_frontend.py`,  `pytest tests/python/test_schedule_ops.py` and python benchmarks are legacy.

Comment (mdavis36) [reply python/python_frontend/fusion_cache.cpp:352]: I will remove those

File: tests/python/direct/test_cutlass_nvfp4_gemm.py (hunk 7 above)
Comment (mdavis36) [reply tests/python/direct/test_cutlass_nvfp4_gemm.py:11]: It should, this would allow compute capabilities in the range of [10.0, 12.0)

File: python/python_frontend/fusion_cache.cpp (hunk 1 above)
Comment (mdavis36) [reply python/python_frontend/fusion_cache.cpp:352]: @rdspring1  This is tested in the executable target `bin/test_python_frontend` from [here](https://github.com/NVIDIA/Fuser/blob/a40edec418a8143767788affc36a9cdf237d3b25/python/tests/python_frontend/test_nvfuser_fusion_cache.cpp#L21). Is this test still valid?

File: tests/python/direct_utils/utils.py [hunk 9]
```diff
@@ -28,6 +28,11 @@ def is_pre_blackwell():
     return prop.major < 10
 
 
+def is_pre_blackwell_12():
```
Comment (wujingyue) [inline tests/python/direct_utils/utils.py]: FYI, I just added `microarchitecture_is_pre` which you can replace is_pre_blackwell_12 with. Having both `blackwell` and `12` in the name is redundant.

File: tests/cpp/test_combined_inner_outer_reduction.cpp (hunk 3 above)
Comment (mdavis36) [reply tests/cpp/test_combined_inner_outer_reduction.cpp]: I went back and forth on what to do for this, this test passes for the smaller input cases but after a certain size the test suite will start to fail the `validateHeuristics` check [here](https://github.com/NVIDIA/Fuser/blob/587812b041a7515b39b62510b23a871aa5002060/tests/cpp/test_combined_inner_outer_reduction.cpp#L1070). 

My theory is it's because the scheduler can no longer perform warp specialization. I'm not entirely sure WHY it no longer can, but I assume it has to do with the smaller smem size of compute 12 devices so the scheduler fails to apply the technique? I saw similar behavior with a below test and found dropping the input allowed the test to pass [here](https://github.com/NVIDIA/Fuser/pull/5284/files/587812b041a7515b39b62510b23a871aa5002060?diff=split&w=0#diff-a26d40a98e83c9b3431f0c18cba0b27c0222f763c770d9c80562513294a13c93R1330).

I decided disabling this test outright could be a reasonable change as I didn't want to reduce the range of inputs this test. I could try and calculate a cutoff threshold instead and skip the failing tests for arch 12.

Comment (naoyam) [reply tests/cpp/test_combined_inner_outer_reduction.cpp]: @liqiangxl Could you take a look?

File: tests/python/direct_utils/utils.py (hunk 9 above)
Comment (mdavis36) [reply tests/python/direct_utils/utils.py]: done.

File: tests/python/direct/test_cutlass_gemm.py (hunk 8 above)
Comment (mdavis36) [reply tests/python/direct/test_cutlass_gemm.py]: done

File: tests/python/direct/conftest.py (hunk 6 above)
Comment (mdavis36) [reply tests/python/direct/conftest.py]: done

File: tests/cpp/test_memory.cpp (hunk 2 above)
Comment (mdavis36) [reply tests/cpp/test_memory.cpp]: done

File: python/python_frontend/fusion_cache.cpp (hunk 1 above)
Comment (wujingyue) [reply python/python_frontend/fusion_cache.cpp:352]: Let me smoke test #5342

File: tests/python/direct_utils/utils.py [hunk 10]
```diff
@@ -115,7 +115,7 @@ def create_sdpa_rng_tensors() -> tuple[torch.Tensor, torch.Tensor]:
     return philox_seed, philox_offset
 
 
-def skip_if_global_memory_below_gb(min_gb: int, didx: int=0):
+def skip_if_global_memory_below_gb(min_gb: int, didx: int = 0):
```
Comment (wujingyue) [inline tests/python/direct_utils/utils.py]:
```suggestion
def skip_if_global_memory_below_gb(min_gb: int, gpu_id: int = 0):
```
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: tests/cpp/test_combined_inner_outer_reduction.cpp (hunk 3 above)
Comment (liqiangxl) [reply tests/cpp/test_combined_inner_outer_reduction.cpp]: Hi @mdavis36, you guess is correct. The actual heuristic depends on the hardware shared memory. Therefore, you can revise `validateHeuristics` to allow just one computation warp group on certain devices.

Comment (liqiangxl) [reply tests/cpp/test_combined_inner_outer_reduction.cpp]: We can just skip this check `EXPECT_TRUE(rparams->computation_warp_groups > 1);` on these devices.

Comment (mdavis36) [reply tests/cpp/test_combined_inner_outer_reduction.cpp]: Sounds good thank you @liqiangxl !!

**Comment by mdavis36** [comment]
!test

**Comment by mdavis36** [comment]
!test
